69 separate python packages for epidata and bindings

dabele opened the pull request:
## Changes

- epidata and bindings in separate python packages that can be installed on their own
- packages have a shared namespace (memilio)
- build and test each package independently in CI
- updated readmes (for #94: added a troubleshooting hint for epidata package installation)
- removed some obsolete files from pycode directory (for #96: deleted the create venv script)

closes #69
closes #96 
closes #94

## Merge Request - GuideLine Checklist 

**Guideline** to check code before resolve WIP and approval, respectively.
As many checkboxes as possible should be ticked.

### Checks by code author:
* [x] There is at least one issue associated with the pull request.
* [x] The branch follows the naming conventions as defined in the [git workflow](git-workflow).
* [ ] New code adheres with the [coding guidelines](coding-guidelines)
* [ ] Tests for new functionality has been added
* [x] A local test was succesful
* [ ] There is appropriate **documentation** of your work. (use doxygen style comments)
* [ ] If new third party software is used, did you pay attention to its license? Please remember to add it to the wiki after successful merging.
* [ ] If new mathematical methods or epidemiological terms are used, has the glossary been updated ? Did you provide further documentation ?
 is present or referenced. Please provide your references.
* [ ] The following questions are addressed in the documentation*:  Developers (what did you do?, how can it be maintained?), For users (how to use your work?), For admins (how to install and configure your work?)
* [x] For documentation: Please write or update the Readme in the current working directory!

### Checks by code reviewer(s):
* [ ] Is the code clean of development artifacts e.g., unnecessary comments, prints, ...
* [ ] The ticket goals for each associated issue are reached or problems are clearly addressed (i.e., a new issue was introduced).
* [ ] There are appropriate **unit tests** and they pass.
* [ ] The git history is clean and linearized for the merge request.
* [ ] Coverage report for new code is acceptable.

KPunktRPunkt commented on pycode/epidata/setup.py:
```diff
@@ -0,0 +1,78 @@
+    packages=find_packages(where = os.path.dirname(os.path.abspath(__file__))),
+    long_description='',
+    test_suite='memilio.epidata_test',
+    install_requires= [
```
It is working with:  python3 -m pip install -e .[pylint]

I have to admit this should be in the readme.

KPunktRPunkt commented on pycode/README.rst:
```diff
@@ -1,110 +1,36 @@
-- freezegun (freezes the time. Thus, the call today can be changed to a specific date.)
-
-To run the tests make 
+If you are a developer, use `python setup.py develop` instead when you are working on the code. Each package contains a test suite that can be run with
```
You have quotes ('') here but in the Readmea it is shown as normal text. There should be some highlighting.

dabele replied:
I mixed up `rst` and `md` format there. Would be easier if all READMEs were the same format.

KPunktRPunkt commented on pycode/memilio-epidata/setup.py:
```diff
@@ -0,0 +1,78 @@
+        'openpyxl',
+        'xlrd', 
+        'requests'
+    ],
```
```suggestion
    ],
    extras_require={"pylint": ["pylint", "pylint_json2html"],
                              "test": ["pyfakefs==4.1.0",
                                         "freezegun",
                                         "coverage"]},
```

dabele replied:
with this, there are two ways to install dev requirements: `pip install .[test]` or `pip install -r requirements-dev.txt`.
There should only be one way (or at least, the requirements should be only listed in one place).

I propose this:
- one extras-require `dev` which combines pylint and test. every developer needs both anyway as far as I can tell (or at least they should) and maybe sometimes installing a few unnecessary packages is better than two separate commands
- delete requirements-dev.txt, it's obsolete if requirements are all listed in setup.py

KPunktRPunkt commented on pycode/memilio-epidata/README.rst:
```diff
@@ -0,0 +1,343 @@
+
+- pyfakefs (creates fake directory to test that expected folders are created and data is written)
+- freezegun (freezes the time. Thus, the call today can be changed to a specific date.)
+
```
```suggestion

This can be done manually using e.g. pip or by
.. code:: sh

          python3 -m pip install -e .[test]
```

KPunktRPunkt commented on pycode/memilio-epidata/README.rst:
```diff
@@ -0,0 +1,343 @@
+* pylint
+* pylint-json2html
+
+After installing the package, run
```
```suggestion
After installing the package manually or using 

.. code:: sh

                    python3 -m pip install -e .[pylint]

run
```

KPunktRPunkt commented:
> I still don't know what to do with the tools directory. I moved it to epidata for now because the only script in there is data related. If I see a directory "tools" in a software repository, I assume it's something used for maintenance of the repository or project somehow, not part of the functionality.

I think where it is, the tool folder is fine. The content of the tool folder is not part of the functionality, it just uses it to generate test data. However, since for tests in the future fictional data should be used, it would be also fine to delete it.

dabele commented:
re-added extras-require to setup.py, but only one category (dev) instead of two (pylint, test). I think all developers should need both linting and testing to verify their work.

but did not remove requirements-dev.txt so far, it's used in CI python test jobs to install only the dependencies without the package itself (the package is loaded as an artifact from the python build job) and haven't found a good replacement without changing all the python jobs. Not worth it, imho.

Added lines about developer install (`pip install .[dev]`) to READMEs. Since the extras-require can only be used when installing with pip, I changed all the READMEs to recommend pip instead of using setup.py directly. As far as I know (I'm no python package expert), this is the recommended way anyway.

KPunktRPunkt commented:
Like the "dev"-related changes. Installation via dev works for both python packages and tests run through. Two small remarks which can be maybe moved to other issues

- [ ] Example folders are mssing Readmes which explain what the scripts do and what input us required
- [ ] AttributeError: module 'memilio.simulation.secir' has no attribute 'SecirModel1' for running migration.py and migration_parameter_study.py I wonder if installation does not work properly althouh it seems so or if something is wrong in the example files.

dabele commented:
> Example folders are mssing Readmes which explain what the scripts do and what input us required

created #136

> AttributeError: module 'memilio.simulation.secir' has no attribute 'SecirModel1' for running migration.py and migration_parameter_study.py I wonder if installation does not work properly althouh it seems so or if something is wrong in the example files.

There already is a separate issue for that (#115), the problems are unrelated to this PR.
